designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
+
+
+Custom Build Commands
+~~~~~~~~~~~~~~~~~~~~~
+
+Users can provide custom build command for each runtime via ``.samrc`` that will be run instead of the default build
+action provided by SAM CLI. SAM CLI will invoke this command for every resource that uses this runtime. This can be any
+arbitrary program that is accessible (via ``PATH`` resolution). The program will be invoked by setting three
+environment variables:
+
+#. ``__SOURCE_DIR`` - Folder where the source files are located
+#. ``__BUILD_DIR`` - Folder where the built artifacts should be written to
+#. ``__RUNTIME`` - AWS Lambda function runtime
+#. ``__TEMPLATE`` - Path to SAM Template
+#. Current Working Directory - Same directory where ``sam build`` command was invoked from
+
+This command is supposed to return an exit code of ``zero`` to indicate success. Non-zero exit codes indicate failure.
+
+Building Native Binaries
+~~~~~~~~~~~~~~~~~~~~~~~~
+To build native binaries, we need to run on an architecture and operating system that is similar to AWS Lambda. We use
+the Docker containers provided by `Docker Lambda <https://github.com/lambci/docker-lambda>`_ project to run the same
+set of commands described above on this container. We will mount source code folder and build folder into the container
+so the commands have access to necessary files.
+
+If the user specified a custom build command, we will attempt to run the command as-is within the Docker Container.
+But it is possible that the command might not be available inside the container.
+
+``.samrc`` Changes
+------------------
+*Explain the new configuration entries, if any, you want to add to .samrc*
+
+We will add a new section to ``.samrc`` where customers can provide custom build actions. This section will look like:
+
+.. code-block:: json
+
+    {
+        "build": {
+            "actions": {
+                "java8": "gradle build",
+                "dotnetcore2.1": "./build.sh"
+            }
+        }
+    }
+
+Security
+--------
+
+*Tip: How does this change impact security? Answer the following questions to help answer this question better:*
+
+**What new dependencies (libraries/cli) does this change require?**
+
+**What other Docker container images are you using?**
+
+**Are you creating a new HTTP endpoint? If so explain how it will be created & used**
+
+**Are you connecting to a remote API? If so explain how is this connection secured**
+
+**Are you reading/writing to a temporary folder? If so, what is this used for and when do you clean up?**
+
+**How do you validate new .samrc configuration?**
+
+
+Documentation Changes
+---------------------
+TBD
+
+Open Questions
+--------------
+
+#. Should we support ``artifacts.json`` now to be future-proof?
+#. Should we create the default ``build`` folder within a ``.sam`` folder inside the project to provide a home for
+   other scratch files if necessary?
+
+
+Task Breakdown
```

Comment:
Perhaps we should also have `sam init` as a task to update existing samples with `.samrc` and `sam build` as part of the README generated

Reply:
`init` samples shouldn't need a custom `.samrc`. Whatever comes built-in with SAM CLI by default should just work with init samples. 

+1 on adding `sam build` command as part of readme. Will add that task to the task list